### PR TITLE
Fix #205 wrap read/write_text in try with raise from

### LINKED ISSUE #205: RFC: pykern.quest
A Quest is "a long or arduous search for something." It is also a Pythonesque word so appropriate. The reason it is called Quest is that it is longer than a simple function call, since it must validate its inputs, e.g. an HTTP request, command line execution (pkcli), a Web Socket message, etc.

There are several classes in sirepo.quest:

* Spec is the decorator that describes the functions inputs and other prerequisites (e.g. permissions)
* Call is instantiated to make a call to a quest function
* Args is a PKDict subclass that holds the inputs
* Result is the return value or reply

Sirepo will subclass these, e.g. a sirepo.api.Base is a subclass of pykern.quest.Call, and used as follows:

```py
    @api_perm.require_user
    def api_deleteSimulation(self):
        req = self.parse_post(id=True)
        simulation_db.delete_simulation(req.type, req.id)
        return self.reply_ok()
```

Would turn into:

```py
    @sirepo.api.Spec('require_user', id='SimId')
    def api_deleteSimulation(self):
        simulation_db.delete_simulation(self.args.type, self.args.id)
        self.done(status='ok')
```

`self.result(status='ok')` would be shortened to `self.result_ok()`.

In the above example, sirepo.api.Spec always adds `type='SimType'` and applies the `@api_perm.require_user`. The value `auth_uid` is also in args. All values in args are validated, and collisions are managed safely, e.g. if a caller tries to supply auth_uid, it will be rejected, since that type of arg is special (the structures for this have to be defined clearly).

args will be parsed from whatever inputs make sense for the way the Call is made. If it is an API request and there is a URI with parameters, those values will be extracted, validated, and add to args. If there is a POST or PUT, those values will be validated and added to args. This is already being done to a degree with parse_post, but this formalizes all inputs.

The types of the inputs are found in pykern.pktype. For Sirepo sirepo.pktype will define its additional types. The type interface has yet to be defined, but this is part of the project.

### PR DESCRIPTION
**Pull request by robnagler**
- [x] Require test?
   - Yes
- [x] Security implications?
    - None
- [ ] UI changes reviewed?
    - No, probably should be
- [ ] Backwards compatible?
    - Yes
- [x] Require doc changes?
    - n/a
- [x] Add [DesignHints](https://github.com/radiasoft/pykern/wiki/DesignHints)?
    - n/a

### CODE REVIEW

File: pykern/pkio.py [hunk 1]
```diff
@@ -336,6 +341,11 @@ def write_text(path, contents):
     from pykern import pkcompat
 
     fn = py_path(path)
```
Comment (mkeilman) [inline pykern/pkio.py:343]: `write_binary()` uses `p = py_path(path)` but this uses `fn = py_path(path)`. A minor point but consistency is helpful

**Review by mkeilman — COMMENTED** [review]
In addition to the naming comment, a few questions:

- One merge comment says this fixes #205 when it should be #227
- Should read/write_binary also get this treatment?
- I'm not sure what the issue is meant to address. The issue has no comment

File: pykern/pkio.py (hunk 1 above)
Comment (robnagler) [reply pykern/pkio.py]: Fix 54877fc

**Comment by robnagler** [comment]
@mkeilman I made a significant change by adding _exception_reason() which is more robust.  Please do a thorough re-review

**Review by mkeilman — APPROVED** [review]
Some notes, but nothing of concern

File: pykern/pkio.py [hunk 2]
```diff
@@ -335,7 +339,28 @@ def write_text(path, contents):
...
+        with io.open(str(p), "wt", encoding=TEXT_ENCODING) as f:
+            f.write(pkcompat.from_bytes(contents))
+    except Exception as e:
+        _exception_reason(e, f"path={path}")
+        raise
+    return p
+
+
+def _exception_reason(exc, reason):
+    reason = '; ' + reason
```
[the remark this answers is not in the document]
Comment (robnagler) [reply pykern/pkio.py]: The checks failed, sorry. You don't have to comment about those. We shouldn't merge if checks fail.

File: pykern/pkio.py [hunk 3]
```diff
@@ -335,7 +339,28 @@ def write_text(path, contents):
...
+def _exception_reason(exc, reason):
+    reason = "; " + reason
+    if hasattr(exc, "reason"):
+        exc.reason += reason
+    if hasattr(exc, "args"):
+        if exc.args is None:
+            exc.args = tuple()
+        if isinstance(exc.args, (tuple, list)):
+            if len(exc.args) == 0:
+                exc.args = (reason,)
```
[the remark this answers is not in the document]
Comment (robnagler) [reply pykern/pkio.py:361]: Yes, that's an interesting problem. I'll change that.

File: pykern/pkio.py [hunk 4]
```diff
@@ -335,7 +339,28 @@ def write_text(path, contents):
...
+        raise
+    return p
+
+
+def _exception_reason(exc, reason):
+    reason = "; " + reason
+    if hasattr(exc, "reason"):
+        exc.reason += reason
+    if hasattr(exc, "args"):
+        if exc.args is None:
```
[the remark this answers is not in the document]
Comment (mkeilman) [reply pykern/pkio.py:357]: 👍

File: pykern/pkio.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (robnagler) [reply pykern/pkio.py:363]: 95ca507

File: pykern/pkio.py [hunk 5]
```diff
@@ -335,7 +339,28 @@ def write_text(path, contents):
...
+            exc.args = tuple()
+        if isinstance(exc.args, (tuple, list)):
+            if len(exc.args) == 0:
+                exc.args = (reason,)
+            elif isinstance(exc.args[0], str):
```
[the remark this answers is not in the document]
Comment (robnagler) [reply pykern/pkio.py:364]: Defensive programming. I wrote a test. 95ca507

File: pykern/pkio.py [hunk 6]
```diff
@@ -335,7 +339,28 @@ def write_text(path, contents):
...
+                x = list(exc.args)
+                x[0] += reason
+                exc.args = tuple(x)
+    # Other cases?
```
[the remark this answers is not in the document]
Comment (robnagler) [reply pykern/pkio.py:368]: Yea, eventually _exception_reason() will move elsewhere and contain more cases.

**Comment by robnagler** [comment]
Fix #227

**Pull request merged**